### PR TITLE
Dataset configuration refactoring

### PR DESCRIPTION
**Pull request by erksch**
With this PR, configuring different (and new) datasets is easier.
There is a new section in the .ini config file called `[dataset]`.
You can specify one of in `datasets.py` predefined datasets by their id and add dataset specific configration by adding a prefix with the dataset name:
```
[dataset]
dataset=simple
dataset_path=./dataset
``` 

To configure a new dataset, make a class in `datasets.py` that inherits the `Dataset` class and implements a `load()` method which sets the `self.image_paths` and `self.bounding_boxes` properties.

```python
class MyDataset(Dataset):
    id = 'my'

    def load(self):
        self.image_paths = ...
        self.bounding_boxes = ....
```

This PR already contains dataset classes for the simple and sign dataset which can be configured with the id `simple`, `sign` or `synthtext`.

### CODE REVIEW

File: example.ini [hunk 1]
```diff
@@ -1,7 +1,9 @@
+[dataset]
+dataset=simple
+simple_dataset_image_locations=../dataset-generator/image_locations.txt
```
Comment (janukobytsch) [inline example.ini]: Would be nice if this was encapsulated by the dataset & you only need to to provide the name (+path) of the dataset

Comment (janukobytsch) [reply example.ini]: Maybe we can talk to Christian to host the datasets on some server & provide them as package so you don't even need to set a local path

File: evaluate.py [hunk 2]
```diff
@@ -37,65 +38,15 @@ def add_path(path):
...
-        return cls(absolute_paths, bboxes)
-
-    def get(self, idx, as_image=True):
-        image = self.image_paths[idx]
-        if as_image:
-            image = Image.open(image)
-        true_bboxes = self.true_bboxes[idx]
-        return image, true_bboxes
-
-    def get_image_name(self, idx):
```
Comment (janukobytsch) [inline evaluate.py:76]: Eval script will break if new dataset doesn't provide these methods

File: example.ini [hunk 3]
```diff
@@ -23,7 +25,7 @@ start_epsilon=1.0
 end_epsilon=0.1
 decay_steps=300000
 # training
-steps=100000
+steps=100
```
Comment (janukobytsch) [inline example.ini]: Why 100?

File: datasets.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        self.image_paths = [os.path.join(self.dataset_path, image_path) for image_path in relative_image_paths]
+        
+        bounding_boxes_file = os.path.join(self.dataset_path, 'bounding_boxes.npy')
+        self.bounding_boxes = np.load(bounding_boxes_file, allow_pickle=True)
+
+class SignDataset(Dataset):
+    id = 'sign'
+
+    def load(self):
+        config_file_path = os.path.join(self.dataset_path, 'training.json')
```
Comment (janukobytsch) [inline datasets.py:56]: How would this work if I want to evaluate on the validation split? (separate `validation.json`)

Comment (janukobytsch) [reply datasets.py:56]: Maybe if there's a predefined split (both `train.json` and `validation.json`), we can load & store them as sub-datasets (composite pattern). If there are no predefined splits, we could just have a train split or generate the splits ourselves.

Comment (janukobytsch) [reply datasets.py:56]: The evaluation script would then be able to evaluate on each of the splits / sub-datasets so you can get both train & test/val metrics.

File: datasets.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
+import json
+import os
+import numpy as np
+import scipy.io as sio
+from PIL import Image
+from pathlib import Path
+
+class Dataset:
+    id = 'noop'
```
Comment (janukobytsch) [inline datasets.py]: Just a hint: Python convention is to use `NotImplemented` / `raise NotImplementedError` for methods/properties that should be overriden by subclasses. There's also an `abc` package for abstract decorators if the dataset class should indeed be abstract.

File: evaluate.py [hunk 6]
```diff
@@ -147,12 +98,13 @@ def f1(precision, recall):
 def main(eval_dirname='evaluations', viz_dirname='episodes', images_dirname='images', plots_dirname='plots', max_sample_size=100):
     print_config()
 
-    if CONFIG['jsonfile_path']:
-        dataset = Dataset.from_json(CONFIG['jsonfile_path'])
-    else:
-        dataset = Dataset.from_numpy(CONFIG['imagefile_path'], CONFIG['boxfile_path'])
+    dataset_id = CONFIG['dataset']
+    dataset = get_dataset(dataset_id)(CONFIG['dataset_path'])
+    dataset.load()
```
Comment (janukobytsch) [inline evaluate.py]: API seems a bit clunky, why not simply `load_dataset(CONFIG['dataset'], CONFIG['dataset_path'])`? :D

File: example.ini (hunk 3 above)
Comment (erksch) [reply example.ini]: Ups that was just for testing.

File: datasets.py (hunk 5 above)
Comment (erksch) [reply datasets.py]: Okay nice thank you, I'll will change it.

File: evaluate.py (hunk 6 above)
Comment (erksch) [reply evaluate.py]: Nice idea!

**Pull request merged**